### PR TITLE
The variable ${workspaceRoot} points at an erroneous location for the mutli-root workspace folders.

### PR DESCRIPTION
**Pull request by stuartZhang**
In my use case, 

(1) There are FOUR sub-projects in the workspace: "list (vue)", "design (vue)", "mobile (vue)"  and "java (server)". 
(2) The ".vscode/settings.json" per project is disabled automatically. Only the ".code-workspace" file for the whole workspace takes effect. That's the behavior of the latest VSCode.
(3) Finally, ${workspaceRoot} always points to the first sub project "list (vue)".

My job is to fix the bug arising in #3.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -7,7 +7,10 @@
 			"type": "extensionHost",
 			"request": "launch",
 			"runtimeExecutable": "${execPath}",
-			"args": ["--extensionDevelopmentPath=${workspaceRoot}" ],
+			"args": [
+				"--disable-extensions",
```
Comment (stuartZhang) [inline .vscode/launch.json]: To enhance the debugging efficiency, the CLI parameter is used to turn off all other VSCode extensions and only the extension "flow-for-vscode" will be run in the [Extension Host] instance.

File: package.json [hunk 2]
```diff
@@ -94,6 +94,7 @@
   "scripts": {
     "vscode:prepublish": "babel ./lib --out-dir=./build --source-maps",
     "compile": "babel ./lib --out-dir=./build --source-maps --watch",
+    "manual_compile": "babel ./lib --out-dir=./build --source-maps",
```
Comment (stuartZhang) [inline package.json]: My machine is terrible and doesn't afford to the watch-mode ES6 compilation.

File: .vscode/tasks.json [hunk 3]
```diff
@@ -9,19 +9,23 @@
...
-	"isBackground": true
+	"command": "npm", // we want to run npm
+	"isShellCommand": true, // the command is a shell script
+	"suppressTaskName": true,
+	"tasks": [{
+		"taskName": "watch",
+		// show the output window only if unrecognized errors occur.
+		"showOutput": "silent",
+		// we run the custom script "compile" as defined in package.json
+		"args": ["run", "compile", "--loglevel", "silent"],
```
Comment (stuartZhang) [inline .vscode/tasks.json]: Your compilation task has been preserved.

File: .vscode/tasks.json [hunk 4]
```diff
@@ -9,19 +9,23 @@
...
+		// Whether the babel compiler is started in watching mode
+		"isBackground": true
+	}, {
+		"taskName": "compile",
+		// show the output window only if unrecognized errors occur.
+		"showOutput": "always",
+		// we run the custom script "compile" as defined in package.json
+		"args": ["run", "manual_compile"],
```
Comment (stuartZhang) [inline .vscode/tasks.json]: This is my manual-compilation task which is configured as a default build command.

File: lib/pkg/flow-base/lib/FlowHelpers.js [hunk 5]
```diff
@@ -154,38 +156,77 @@ async function canFindFlow(flowPath: string): Promise<boolean> {
...
-    } else {
-      const extensionRoot = path.resolve(__dirname, '../../../../');
-      global.cachedPathToFlowBin = nodeModuleFlowLocation(extensionRoot)
-    }
-
-    logger.info("Path to Flow: " + global.cachedPathToFlowBin);
-   }
-
-  return global.cachedPathToFlowBin
+    global.cachedPathToFlowBin[workspaceRoot] = (async () => {
```
Comment (stuartZhang) [inline lib/pkg/flow-base/lib/FlowHelpers.js]: By means of storing the promise in the cache variable (i.e. cachedPathToFlowBin), resist the data race.

File: lib/pkg/flow-base/lib/FlowHelpers.js [hunk 6]
```diff
@@ -154,38 +156,77 @@ async function canFindFlow(flowPath: string): Promise<boolean> {
...
+    logger.info("workspaceRoot: " + workspaceRoot);
     const config = global.vscode.workspace.getConfiguration('flow');
     const shouldUseNodeModule = config.get('useNPMPackagedFlow');
-    const userPath = config.get('pathToFlow')
-      .replace('${workspaceRoot}', workspaceRoot);
+    const userPath = config.get('pathToFlow').replace('${workspaceRoot}', workspaceRoot);
     const nodeModuleFlowPath = nodeModuleFlowLocation(workspaceRoot);
-
-    if (shouldUseNodeModule && await canFindFlow(nodeModuleFlowPath)){
-      global.cachedPathToFlowBin = nodeModuleFlowPath;
```
Comment (stuartZhang) [inline lib/pkg/flow-base/lib/FlowHelpers.js]: Due to the lingering assignment to the cache variable "cachedPathToFlowBin", the data race causes that the path computation is executed multiple times.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+
+### 0.7.7
+
+* Serialize the flow-server operations (including startup and ping) not to launch too many flow-server processes due to the data race.
+* Turn on more logs for the development mode.
+
+### 0.7.5
+
+* Make the variable `${workspaceRoot}` compatible with the VSCode workspace folders.
+
```
Comment (orta) [inline CHANGELOG.md]: We've not made a release, so all of those versions are still not used in this project

File: lib/flowCoverage.js [hunk 8]
```diff
@@ -26,7 +26,7 @@ export class Coverage {
   state: State;
 
   static createStatusBarItem() {
-    const coverageStatus = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
+    const coverageStatus = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 10);
```
Comment (orta) [inline lib/flowCoverage.js]: What is the 10 for?

File: lib/flowMain.js [hunk 9]
```diff
@@ -35,7 +35,7 @@ export function activate(context:ExtensionContext): void {
 
 	setupLogging()
 	checkNode()
-	checkFlow()
+	// checkFlow()
```
Comment (orta) [inline lib/flowMain.js]: Is this temporary?

File: lib/pkg/flow-base/lib/FlowHelpers.js [hunk 10]
```diff
@@ -154,38 +156,82 @@ async function canFindFlow(flowPath: string): Promise<boolean> {
...
+    activeDoc = textDocuments[textDocuments.length - 1];
+  }
+  if (activeDoc && (workspaceFolder = workspaceFolders.find(findActFolder(activeDoc)))) {
+    workspaceRoot = workspaceFolder.uri.fsPath;
+  } else {
+    workspaceRoot = rootPath;
+  }
+  return workspaceRoot;
+}
+
```
Comment (orta) [inline lib/pkg/flow-base/lib/FlowHelpers.js]: I don't understand why this is referencing css?

File: package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (orta) [inline package.json]: Please don't add a package.lock to this project, it already has a lock file

**Comment by orta** [comment]
I'm going to close this, as I don't think you have any plans to move any of it upstream, thanks for at least giving it a shot at the start.

**Pull request closed without merging**